ffi/src/ATen/Cast.hs

```diff
@@ -36,155 +36,7 @@ import Foreign.Ptr
 
-instance Castable () () where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CBool CBool where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CChar CChar where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CClock CClock where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CDouble CDouble where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CFile CFile where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CFloat CFloat where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CFpos CFpos where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CInt CInt where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CIntMax CIntMax where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CIntPtr CIntPtr where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CJmpBuf CJmpBuf where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CLLong CLLong where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CLong CLong where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CPtrdiff CPtrdiff where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CSChar CSChar where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CSUSeconds CSUSeconds where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CShort CShort where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CSigAtomic CSigAtomic where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CSize CSize where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CTime CTime where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CUChar CUChar where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CUInt CUInt where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CUIntMax CUIntMax where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CUIntPtr CUIntPtr where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CULLong CULLong where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CULong CULong where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CUSeconds CUSeconds where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CUShort CUShort where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable CWchar CWchar where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable Int8 Int8 where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable Int16 Int16 where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable Int32 Int32 where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable Int64 Int64 where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable Word8 Word8 where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable Word16 Word16 where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable Word32 Word32 where
-  cast x f = f x
-  uncast x f = f x
-
-instance Castable Word64 Word64 where
+instance Castable a a where
```

Comment:
Don't you have to add `{-# OVERLAPPABLE #-}` here?

Reply:
Seems like it works out nicely! There's only one extra self-cast I had to add some time ago that you missed and that's causing the compile error.

Reply:
When we want to overlap self-cast of a instance, we have to add `{-# OVERLAPPABLE #-}`.
For now, I feel the overlap-pragma is not needed.